metrics/graphite/graphite.go

```diff
@@ -0,0 +1,345 @@
+// windowed HDR histogram which drops data older than five minutes.
+//
+// The histogram exposes metrics for each passed quantile as gauges. Quantiles
+// should be integers in the range 1..99. The gauge names are assigned by
+// using the passed name as a prefix and appending "_pNN" e.g. "_p50".
+//
+// The values of this histogram will be periodically emitted in a Graphite-compatible
+// format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewHistogram(name string, minValue, maxValue int64, sigfigs int, quantiles ...int) (metrics.Histogram, error) {
+	gauges := map[int]metrics.Gauge{}
+	for _, quantile := range quantiles {
+		if quantile <= 0 || quantile >= 100 {
+			return nil, fmt.Errorf("invalid quantile %d", quantile)
+		}
+		gauges[quantile] = e.gauge(fmt.Sprintf("%s_p%02d", name, quantile))
+	}
+	h := newWindowedHistogram(name, minValue, maxValue, sigfigs, gauges, e.logger)
+
+	e.mtx.Lock()
+	e.histograms = append(e.histograms, h)
+	e.mtx.Unlock()
+	return h, nil
+}
+
+// NewTimeHistogram returns a TimeHistogram wrapper around the windowed
+// HDR histrogram provided by this package.
+func (e *emitter) NewTimeHistogram(name string, unit time.Duration, minValue, maxValue int64, sigfigs int, quantiles ...int) (metrics.TimeHistogram, error) {
+	h, err := e.NewHistogram(name, minValue, maxValue, sigfigs, quantiles...)
+	if err != nil {
+		return nil, err
+	}
+	return metrics.NewTimeHistogram(unit, h), nil
+}
+
+// NewGauge returns a Gauge whose value will be periodically emitted in
+// a Graphite-compatible format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewGauge(name string) metrics.Gauge {
+	e.mtx.Lock()
+	defer e.mtx.Unlock()
+	return e.gauge(name)
+}
+
+func (e *emitter) gauge(name string) metrics.Gauge {
+	g := &gauge{name, 0}
+	e.gauges = append(e.gauges, g)
+	return g
+}
+
+func (e *emitter) dial() error {
+	if e.tcp {
+		tAddr, err := net.ResolveTCPAddr("tcp", e.addr)
+		if err != nil {
+			return err
+		}
+		e.conn, err = net.DialTCP("tcp", nil, tAddr)
+		if err != nil {
+			return err
+		}
+	} else {
+		uAddr, err := net.ResolveUDPAddr("udp", e.addr)
+		if err != nil {
+			return err
+		}
+		e.conn, err = net.DialUDP("udp", nil, uAddr)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Start will kick off a background goroutine to
+// call Flush once every interval.
+func (e *emitter) Start(interval time.Duration) error {
+	var err error
+	e.start.Do(func() {
+		err = e.dial()
+		if err != nil {
+			return
+		}
+		go func() {
+			t := time.Tick(interval)
+			for {
+				select {
+				case <-t:
+					e.Flush()
+				case <-e.stop:
+					return
+				}
+			}
+		}()
+	})
+	return err
+}
+
+// Stop will flush the current metrics and close the
+// current Graphite connection, if it exists.
+func (e *emitter) Stop() error {
+	if e.conn == nil {
+		return nil
+	}
+	// stop the ticking flush loop
+	e.stop <- true
```

Comment:
Roger that. I've added logic to `nil` the connection after Stop. After the first call, this should just immediately return.